### PR TITLE
NullableResponse<T> and consumption by tables

### LINKED ISSUE #27852: Add `NullableResponse<T>` that `Response<T>` inherits from
`Response<T>` requires that `T Value` is not null.  When we want to work around that and return a `Response<T>` with no `Value` _(like conditional requests that return 3XX, batching requests that haven't been sent to the service yet, DPG errors that were classified as success, ...)_ we usually `throw` when accessed.  See [NoBodyResponse](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs) and [ErrorResponse](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/assets/Generator.Shared/ErrorResponse.cs) as examples.

While that _technically_ meets the contract for `Response<T>`, it doesn't match the spirit of it.  We'd like to introduce a `NullableResponse<T>` that we can start using for this case.  To avoid a bifurcated response hierarchy, we'd like to make the existing `Response<T>` inherit from it.  Here's a rough example of what the API might look like:

```C#
public abstract class NullableResponse<T> {
    public abstract bool HasValue { get; }
    public abstract T Value { get;  }
    public abstract Response GetRawResponse() {}
}
 
public abstract class Response<T> : NullableResponse<T> {
     [EditorBrowsable(EditorBrowsableState.Never)]
     public override bool HasValue => true;
     public static implicit operator T(Response<T> response)
}
```

We also want to evaluate whether DPG protocol methods (and anything accepting `RequestContext`) should return this `NullableResponse<T>` instead of regular `Response<T>`.

### PR DESCRIPTION
**Pull request by christothes**
closes #27852

### CODE REVIEW

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj
Comment (jsquire) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj:17]: I'm not sure that I follow why this is needed.  Did something change that I'm overlooking?

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (jsquire) [inline sdk/core/Azure.Core/CHANGELOG.md:195]:
```suggestion
- Avoid a causing and ignoring an exception when setting network stream timeout on .NET Core.
```

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (jsquire) [inline sdk/core/Azure.Core/src/NullableResponseOfT.cs:41]: Out of curiosity, why expression-bodied methods above and the long form here?

File: sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj
Comment (jsquire) [inline sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj:24]: Not sure that I understand why this is needed.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (jsquire) [inline sdk/tables/Azure.Data.Tables/src/TableClient.cs:703]:
```suggestion
        /// <returns> The <see cref="NullableResponse{T}" whose <c>HasValue</c> property will return <c>true</c> if the entity existed, otherwise <c>false</c>.</returns>
```

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (jsquire) [inline sdk/tables/Azure.Data.Tables/src/TableClient.cs:720]:
```suggestion
        /// <returns> The <see cref="NullableResponse{T}" whose <c>HasValue</c> property will return <c>true</c> if the entity existed, otherwise <c>false</c>.</returns>
```

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (jsquire) [inline sdk/tables/Azure.Data.Tables/src/TableClient.cs:723]: Should we consider opening an issue to tweak this analyzer?   Seems like it may need awareness of the new type.

File: sdk/core/Azure.Core/CHANGELOG.md
Comment (annelo-msft) [inline sdk/core/Azure.Core/CHANGELOG.md:233]: nit: have we made sure IDE settings across the team aren't going to cause whitespace edit wars across changes?

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/NullableResponseOfT.cs:43]: Should this check Value for null?

File: sdk/tables/Azure.Data.Tables/CHANGELOG.md
Comment (annelo-msft) [inline sdk/tables/Azure.Data.Tables/CHANGELOG.md:9]:
```suggestion
- `TableClient.GetEntityIfExists` now returns `NullableResponse<T>` which has a `HasValue` property which returns false when the entity did not exist. Accessing the `Value` property in this case will throw an exception.
```

File: sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj
Comment (jsquire) [reply sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj:24]: But there's no explicit use of the new features/types here.   That's why I'm asking.  😄

File: sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs
Comment (annelo-msft) [inline sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs:40]: Isn't this an unacceptable breaking change, per [D.11.6](https://learning.oreilly.com/library/view/framework-design-guidelines/9780135896457/app04.xhtml#secD_11)?

I think we may be stuck with the existing method the way it is, but we can follow the new guidance for new methods going forward.  Or maybe architects would approve this breaking change this once.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (annelo-msft) [inline sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: I would have expected to see us return an internal type that derived directly from NullableResponse<T> here?  It feels awkward that we return a type that hides and then reinstates "HasValue".

**Review by annelo-msft — COMMENTED** [review]
We should add tests to Core for the new type.

File: sdk/core/Azure.Core/src/ResponseOfT.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/ResponseOfT.cs:21]: nit: Update refdocs to refer to HasValue (or use inheritdoc)

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj
Comment (christothes) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj:17]: Users of `NoResponseBody` need to take these changes from Core for it to compile.

File: sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj
Comment (christothes) [reply sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj:24]: Same comment as above.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:723]: https://github.com/Azure/azure-sdk-tools/issues/4444

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj
Comment (jsquire) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj:17]: Ah, gotcha.   Indirect use due to the new hierarchy.

File: sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs:40]: We haven't GA'd the current method yet.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: I'm not sure it's worth adding a new type unless the API behavior would be different. Is there something that adding a dedicated type here would add that I'm missing?

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:703]: Thanks - I couldn't remember what the generic syntax was.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (annelo-msft) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: It's what it would subtract.  I could cast this to Response<T> or ask `response is Response<T>` and reasonably assume I'd gotten a value for T, since that's what the semantics of Response<T> imply, and it hides `HasValue`.  If we're going to return a Response<T> when T is null, it feels like we're undoing all the work we've done to add a type that indicates the T could be null.  But I'm fine if you wanted to get an architect perspective to see if I'm being too nit-picky.

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/NullableResponseOfT.cs]: I was more worried about `Value` - can't this throw if `HasValue` is false?

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: While it seems like an edge case, it's a good point that someone could cast this to `Response<T>` and get bad hints. 

Would you name the internal type `NullResponse<T>`, `EmptyResponse<T>`, `NoValueResponse<T>`, (any other ideas)?

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs]: Agreed - it's covered in the latest changes.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (jsquire) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: What about `NoContentResponse<T>` - that would seem to align with the official nomenclature for an HTTP 204

Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: After introducing this new internal response type, I think `NoBodyResponse<T>` should effectively be deprecated moving forward. 

Public APIs that already return it are stuck (it's only a hand full of APIs that use it today), but going forward, it would have the same problem as called out here.

Comment (annelo-msft) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: Cool.  I was going to vote for `NullResponse<T>`, to match the corresponding `ValueResponse<T>` and parent, but I don't feel super strongly since it's an internal API.

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs]: Not sure how I missed that.

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: Why is this preferable to just leaving it as empty string?

Comment (christothes) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: I think it is more explicit - similar to "this page intentionally blank".

Comment (annelo-msft) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: Another convention we have in the code base is to return `"<null>"` ... if this is sanctioned, it might read nicely in the NullableResponse type

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: Yeah I would stick with either empty string or "<null>". If users need to know if the value is really null, they have to check HasValue anyway so it doesn't help much to try to make the string representation convey that.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: For that matter, should we include HasValue in the ToString here, and add back the override of ToString into ResponseOfT for back compat?

Comment (christothes) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: I think the value of HasValue is implied by this message. What's the back-compat scenario you had in mind? To restore the throwing behavior, we had before?

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (JoshLove-msft) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: Seems like we are just trying to represent a NullableResponse that is, in fact, null. So NullResponse seems to convey that. We should probably seal this type so nobody tries to change the semantics.

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: No, I meant that if we include HasValue in the ToString for this type, we probably would want to NOT include it for ResponseOfT to retain the existing format.

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: My current vote is `NoValueResponse<T>` since "nullable" is a misnomer we only used due to the similarity with the Nullable<T> API. But in implementation, what we are needing to do is create counterpart to `ValueResponse<T>`

Comment (JoshLove-msft) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:789]: That makes sense to me.

File: sdk/core/Azure.Core/src/NullableResponseOfT.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: But the format is effectively un-changed. `Response<T>` must always have a value.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/NullableResponseOfT.cs:16]: Yes, agreed. I'm saying if we were to add HasValue into the ToString of NullableResponseOfT, we would not want to include it in the ToString of ResponseOfT.

File: sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs:19]: Please add a comment saying this should not be used for new code.

Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs:19]: I was going to do that, but I think it would be more effective to tweak the analyzer rule instead.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs:19]: I didn't realize that we did this.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs:19]: Since we've filed an issue and haven't addressed it yet, should we at least add a comment as a safeguard?

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/NoBodyResponseOfT.cs:22]: added

File: sdk/core/Azure.Core/src/Shared/NoValueResponseOfT.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/NoValueResponseOfT.cs:42]: This name seems to correspond to the `NoBodyResponse` above ... it's possible that a `NullableResponse<T>` could have a body, but no value (i.e. if the payload failed to deserialize, or if the body contained an error).  It seems like another name might be better, or would it make sense to return an InvalidOperationException, since that's what `Nullable<T>` does?

File: sdk/core/Azure.Core/tests/NullableResponseOfTTests.cs
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/NullableResponseOfTTests.cs:29]: Would it make sense to also add tests for `NoValueResponse<T>`?

File: sdk/tables/Azure.Data.Tables/src/TableClient.cs
Comment (annelo-msft) [inline sdk/tables/Azure.Data.Tables/src/TableClient.cs:695]: Could this throw if response.Value is null?
If we know it won't (e.g. it's a `Response<T>`, would it make sense to downcast to `Response<T>` instead so as not to create a new instance?

Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableClient.cs:695]: It will never be null. If the entity did not exist, this would have thrown in `GetEntityInternalAsync`

File: sdk/core/Azure.Core/src/Shared/NoValueResponseOfT.cs
Comment (christothes) [reply sdk/core/Azure.Core/src/Shared/NoValueResponseOfT.cs:42]: > it's possible that a NullableResponse<T> could have a body, but no value (i.e. if the payload failed to deserialize, or if the body contained an error).

I don't think this type is intended to represent those scenarios. We should throw in those examples.

>  It seems like another name might be better, or would it make sense to return an InvalidOperationException, since that's what Nullable<T> does?

I don't feel strongly about the name since it's private, but I like using a known exception type as you suggest.

**Review by jsquire — APPROVED** [review]

**Comment by heaths** [comment]
FWIW, this broke Key Vault's sample, "sharelink". @weshaggard, @sima-zhu, @JoshLove-msft I know you've been making changes with the pipelines, but seems in those changes we no longer build as much - like the Key Vault samples that used to build. Once this change was merged, the `net - keyvault` and `net - keyvault - tests` pipelines started breaking. Maybe autorest updates should still be running all pipelines before merging? Seems we're missing coverage.

**Comment by heaths** [comment]
I opened #32064 to track instead of commenting on this issue.

**Pull request merged**
